### PR TITLE
Handle invalid subscriber cache entries from redis

### PR DESCRIPTION
**Pull request by majchrosoft**
Resolves https://github.com/nuwave/lighthouse/issues/2085

### CODE REVIEW

**Comment by spawnia** [comment]
Please add a failing test case first.

**Comment by majchrosoft** [comment]
@spawnia i've extended test case when unserializa returns false

**Comment by majchrosoft** [comment]
@spawnia sorry for all this commits, but my child was interrupting me and i've done some mistakes ...

File: src/Subscriptions/Storage/RedisStorageManager.php [hunk 1]
```diff
@@ -74,9 +74,11 @@ public function subscribersByTopic(string $topic): Collection
 
         return (new Collection($subscribers))
             ->map(function (?string $subscriber): ?Subscriber {
-                return is_string($subscriber)
-                    ? unserialize($subscriber)
-                    : null;
+                return
+                    is_string($subscriber)
+                    && @unserialize($subscriber) !== false
```
Comment (spawnia) [inline src/Subscriptions/Storage/RedisStorageManager.php:79]: We do not use error suppression through `@`. Also, calling `unserialize` twice is wildly inefficient.

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/Subscriptions/Storage/RedisStorageManager.php [hunk 2]
```diff
@@ -74,9 +74,13 @@ public function subscribersByTopic(string $topic): Collection
 
         return (new Collection($subscribers))
             ->map(function (?string $subscriber): ?Subscriber {
-                return is_string($subscriber)
-                    ? unserialize($subscriber)
-                    : null;
+                $deserializedSubscriber = unserialize($subscriber);
```
Comment (spawnia) [inline src/Subscriptions/Storage/RedisStorageManager.php:77]: Now this is potentially calling `unserialize` with `null`.

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/Subscriptions/Storage/RedisStorageManager.php (hunk 2 above)
Comment (majchrosoft) [reply src/Subscriptions/Storage/RedisStorageManager.php:80]: sure i known it,
just wanted to fast 

right now it's better

**Comment by spawnia** [comment]
`unserialize()` fails with an exception - see https://github.com/nuwave/lighthouse/runs/5534931661?check_suite_focus=true#step:11:41

**Comment by majchrosoft** [comment]
> `unserialize()` fails with an exception - see https://github.com/nuwave/lighthouse/runs/5534931661?check_suite_focus=true#step:11:41

@spawnia I have had exactly the same issue when i was testing locally.
I've already have tried with
```

                $deserializedSubscriber = (function () use ($subscriber) {
                    try {
                        return unserialize($subscriber);
                    } catch (\Throwable $unserializeFailed) {
                        return false;
                    }
                })();
```
but still without success.
The only solution was to suppress with `@`

**Comment by majchrosoft** [comment]
@spawnia 
here i found conversation  about that topic
https://stackoverflow.com/questions/12684871/how-to-catch-unserialize-exception

**Comment by spawnia** [comment]
No idea what setup you were using to run the tests, but it worked fine with the docker setup.

**Comment by majchrosoft** [comment]
> No idea what setup you were using to run the tests, but it worked fine with the docker setup.

You are right. 

I've made mistake and edited another copy of class (the one in my project - which overrides in container the default one).
I 've run test once again without knowing that i did not changed nothing.

Thanks for fast responses

**Pull request merged**